pkg/runtime/serializer/codec_factory.go

```diff
@@ -17,36 +17,43 @@ limitations under the License.
 package serializer
 
 import (
+	"io/ioutil"
+
 	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/runtime/serializer/json"
 	"k8s.io/kubernetes/pkg/runtime/serializer/recognizer"
+	"k8s.io/kubernetes/pkg/runtime/serializer/streaming"
 	"k8s.io/kubernetes/pkg/runtime/serializer/versioning"
 )
 
+// serializerExtensions are for serializers that are conditionally compiled in
+var serializerExtensions = []func(*runtime.Scheme) (serializerType, bool){}
+
 type serializerType struct {
 	AcceptContentTypes []string
 	ContentType        string
 	FileExtensions     []string
-	Serializer         runtime.Serializer
-	PrettySerializer   runtime.Serializer
-}
 
-// NewCodecFactory provides methods for retrieving serializers for the supported wire formats
-// and conversion wrappers to define preferred internal and external versions. In the future,
-// as the internal version is used less, callers may instead use a defaulting serializer and
-// only convert objects which are shared internally (Status, common API machinery).
-// TODO: allow other codecs to be compiled in?
-// TODO: accept a scheme interface
-func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
-	return newCodecFactory(scheme, json.DefaultMetaFactory)
+	Serializer       runtime.Serializer
+	PrettySerializer runtime.Serializer
+	// RawSerializer serializes an object without adding a type wrapper. Some serializers, like JSON
+	// automatically include identifying type information with the JSON. Others, like Protobuf, need
+	// a wrapper object that includes type information. This serializer should be set if the serializer
+	// can serialize / deserialize objects without type info. Note that this serializer is will always
+	// be expected to pass into or a gvk to Decode, since no type information will be available.
+	RawSerializer runtime.Serializer
+
+	// Specialize gives the type the opportunity to return a different serializer implementation if
+	// the content type contains alternate operations. Here it is used to implement "pretty" as an
+	// option to application/json, but could also be used to support strong typing via the content-type
+	// header.
+	Specialize func(map[string]string) (runtime.Serializer, bool)
```

Comment:
Yes, you are correct.  Strong typing might be:

    application/protobuf; type=k8s.io/kubernetes/pkg/api/v1.Pod

where the caller asks for protobuf, specifies the type, and gets back a
message in the exact format.  GRPC does this with a header, but it's
effectively avoiding the need for a runtime type mapper.  I'll try to
clarify the comment to something more useful.

On Tue, Mar 22, 2016 at 4:37 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/runtime/serializer/codec_factory.go
> <https://github.com/kubernetes/kubernetes/pull/23288#discussion_r56951070>
> :
>
> > -func NewCodecFactory(scheme *runtime.Scheme) CodecFactory {
> > -	return newCodecFactory(scheme, json.DefaultMetaFactory)
> > +	Serializer       runtime.Serializer
> > +	PrettySerializer runtime.Serializer
> > +	// RawSerializer serializes an object without adding a type wrapper. Some serializers, like JSON
> > +	// automatically include identifying type information with the JSON. Others, like Protobuf, need
> > +	// a wrapper object that includes type information. This serializer should be set if the serializer
> > +	// can serialize / deserialize objects without type info. Note that this serializer is will always
> > +	// be expected to pass into or a gvk to Decode, since no type information will be available.
> > +	RawSerializer runtime.Serializer
> > +
> > +	// Specialize gives the type the opportunity to return a different serializer implementation if
> > +	// the content type contains alternate operations. Here it is used to implement "pretty" as an
> > +	// option to application/json, but could also be used to support strong typing via the content-type
> > +	// header.
> > +	Specialize func(map[string]string) (runtime.Serializer, bool)
>
> Let me clarify if I understand correctly.
> Based on the ContentType, we choose serializerType, and for a given
> serializerType, we give opportunity to specialize it based on additional
> options. Am I right?
> If so, I don't understand that last sentence of the comment:
> "but could also be used to support strong typing via the content-type
> header"
>
> Can you please clarify?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23288/files/d1795d969a42d04fad220f4c75f2b95f1cfceb85#r56951070>
>